extension/src/controllers/bulk-export-controller.ts

```diff
@@ -0,0 +1,253 @@
+                request.message &&
+                request.message.command === 'card-exported' &&
+                request.src === this._context.video.src
+            ) {
+                const exported = request.message as CardExportedMessage;
+                const isBulk = exported.isBulkExport;
+                if (!isBulk) {
+                    return false;
+                }
+
+                this._inFlight = false;
+                this._currentIndex++;
+                this._lastSkippedDuplicate = !!exported.skippedDuplicate;
+                this._notifyProgress();
+                this._sendNext();
+            }
+
+            return false;
+        };
+
+        browser.runtime.onMessage.addListener(this._listener);
+    }
+
+    unbind() {
+        if (this._listener) {
+            browser.runtime.onMessage.removeListener(this._listener);
+            this._listener = undefined;
+        }
+    }
+
+    async start() {
+        if (this._exporting) {
+            return;
+        }
+
+        const subtitles = this._context.subtitleController.subtitles;
+        if (!subtitles || subtitles.length === 0) {
+            return;
+        }
+
+        // Compute queue: all non-empty subtitle indices
+        const subtitleIndices = subtitles
+            .map((s, i) => ({ s, i }))
+            .filter(({ s }) => s.text !== '')
+            .map(({ i }) => i);
+
+        if (subtitleIndices.length === 0) {
+            return;
+        }
+
+        // Determine start index using the current subtitle
+        const [currentSubtitle] = this._context.subtitleController.currentSubtitle();
+        let startQueueIndex = 0;
+        if (currentSubtitle) {
+            const currentIdx = subtitles.findIndex(
+                (s) => s.start === currentSubtitle.start && s.end === currentSubtitle.end && s.text === currentSubtitle.text
+            );
+            if (currentIdx >= 0) {
+                const found = subtitleIndices.findIndex((i) => i >= currentIdx);
+                startQueueIndex = found >= 0 ? found : 0;
+            }
+        }
+
+        this._queue = startQueueIndex > 0 ? subtitleIndices.slice(startQueueIndex) : subtitleIndices;
+        this._currentIndex = 0;
+        this._exporting = true;
+        this._cancelled = false;
+        this._inFlight = false;
+
+        // Inform background (for CardPublisher state) and any UIs (e.g. SidePanel) about start and total
+        const startedMessage = {
+            sender: 'asbplayerv2',
+            message: {
+                command: 'bulk-export-started',
+                total: this._queue.length,
+            },
+            src: this._context.video.src,
+        };
+        try {
+            await browser.runtime.sendMessage(startedMessage);
+        } catch {}
+
+        // Kick off first item
+        this._sendNext();
+    }
+
+    async cancel() {
+        if (!this._exporting || this._cancelled) {
```

Comment:
I was somehow able to trigger an edge-case where we get into this block even though `SidePanel` is still displaying "in progress." So maybe as a defensive measure you can send the `bulk-export-cancelled` from here as well.